src/activity_item_handling.cpp

```diff
@@ -916,9 +932,26 @@ void activity_on_turn_move_loot( player_activity &, player &p )
         }
 
         auto items = std::vector<item *>();
-        for( auto &it : g->m.i_at( src_loc ) ) {
-            if( !it.made_of_from_type( LIQUID ) ) { // skip unpickable liquid
-                items.push_back( &it );
+
+        //Check source for cargo part
+        //map_stack and vehicle_stack are different types
+        //No way to avoid for loop duplication
```

Comment:
Yes, there is: write a function template (with `vehicle_stack` / `map_stack` being the template arguments).

Reply:
I've never really learned about c++ template stuff properly. I can read them and sortof understand what they're doing but don't know how to write them from scratch to meet a given requirement.

Reply:
I was mostly agitated by the comment itself (being incorrect). You can also use the runtime polymorphism: `map_stack`and `vehicle_stack` inherit from `item_stack` and the later provides all the functions you need here (to iterate over it).

Reply:
I tried declaring an item_stack type because I saw they both inherited from it, but it's a purely virtual class so it wouldn't let me :/

Reply:
You'll need to make it a reference (or pointer) to an `item_stack`.  You can't have an actual `item_stack`, but you can have a reference to a thing that derives from it.

Reply:
All the examples of abstract class polymorphism show how to declare it as a pointer, but doing so makes Gorgon angry, and I have no idea how to declare an item_stack reference without supplying an initializer to allow it to be either type.

Reply:
You have to initialize a reference at declaration time.  Which in this case is going to be a bit tricky.  The neatest way I can think of would be to put the for loop in a lambda taking a `const item_stack&` and call that lambda in the two places.

Reply:
I don't think it can be a const item_stack because the vector it's being shoved into can't hold const items because they're being mutated, but it looks like the lambda worked. Thanks.